<p style="text-align:center">
    <a href="https://skills.network" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="300" alt="Skills Network Logo">
    </a>
</p>

# Importing Dataset

Estimated time needed: **15** minutes

## Objectives

After completing this lab you will be able to:

-   Acquire data in various ways
-   Obtain insights from data with Pandas library


<h2>Table of Contents</h2>

<div class="alert alert-block alert-info" style="margin-top: 20px">
<ol>
    <li><a href="#Data-Acquisition">Data Acquisition</a>
    <li><a href="#Basic-Insights-from-the-Data-set">Basic Insights from the Data set</a></li>
</ol>


</div>
<hr>


# Data Acquisition
<p>
A data set is typically a file containing data stored in one of several formats. Common file formats containing data sets include: .csv, .json, .xlsx etc. The data set can be stored in different places, on your local machine, on a server or a websiite, cloud storage and so on.<br>

To analyse data in a Python notebook, we need to bring the data set into the notebook. In this section, you will learn how to load a data set into our Jupyter Notebook.<br>

In our case, the Automobile Data set is an online source, and it is in a CSV (comma separated value) format. Let's use this data set as an example to practice reading data.
<ul>
    <li>Data source: <a href="https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data" target="_blank">https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data</a></li>
    <li>Data type: csv</li>
</ul>
The Pandas Library is a very popular and very useful tool that enables us to read various datasets into a data frame; our Jupyter notebook platforms have a built-in <b>Pandas Library</b> so that all we need to do is import Pandas without installing.
</p>


In [18]:
# uncomment the lines below if you need to install specific version of libraries if using this notebook 
# in an environment where these libraries are not installed 
# Install specific versions using pip
!pip install pandas==1.3.3
!pip install numpy==1.21.2


  Using cached pandas-1.3.3.tar.gz (4.7 MB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'error'


  error: subprocess-exited-with-error
  
  pip subprocess to install build dependencies did not run successfully.
  exit code: 2
  
  [117 lines of output]
  Ignoring numpy: markers 'python_version == "3.7" and (platform_machine != "arm64" or platform_system != "Darwin") and platform_machine != "aarch64"' don't match your environment
  Ignoring numpy: markers 'python_version == "3.8" and (platform_machine != "arm64" or platform_system != "Darwin") and platform_machine != "aarch64"' don't match your environment
  Ignoring numpy: markers 'python_version == "3.7" and platform_machine == "aarch64"' don't match your environment
  Ignoring numpy: markers 'python_version == "3.8" and platform_machine == "aarch64"' don't match your environment
  Ignoring numpy: markers 'python_version == "3.8" and platform_machine == "arm64" and platform_system == "Darwin"' don't match your environment
  Ignoring numpy: markers 'python_version == "3.9" and platform_machine == "arm64" and platform_system == "Da

In [19]:
# import pandas library
import pandas as pd
import numpy as np

<h2>Read Data</h2>
<p>
We utilize the <code>pandas.read_csv()</code> function for reading CSV files. However, in this version of the lab, which operates on JupyterLite, the dataset needs to be downloaded to the interface using the provided code below.
</p>


The functions below will download the dataset into your browser:


In [22]:
from pyodide.http import pyfetch

async def download(url, filename):
    response = await pyfetch(url)
    if response.status == 200:
        with open(filename, "wb") as f:
            f.write(await response.bytes())

ModuleNotFoundError: No module named 'pyodide'

In [23]:
file_path='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Data%20files/auto.csv'

To obtain the dataset, utilize the download() function as defined above:


In [24]:
await download(file_path, "auto.csv")
file_name="auto.csv"

NameError: name 'download' is not defined

Utilize the Pandas method `read_csv()` to load the data into a dataframe.


In [25]:
df = pd.read_csv(file_name)

NameError: name 'file_name' is not defined

> Note: This version of the lab is working on JupyterLite, which requires the dataset to be downloaded to the interface.While working on the downloaded version of this notebook on their local machines(Jupyter Anaconda), the learners can simply **skip the steps above,** and simply use the URL directly in the `pandas.read_csv()` function. You can uncomment and run the statements in the cell below.


In [28]:
filepath = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Data%20files/auto.csv"
df = pd.read_csv(filepath, header=None)

After reading the data set, we can use the <code>data_frame.head(n)</code> method to check the top n rows of the data frame, where n is an integer. Contrary to <code>data_frame.head(n)</code>, <code>data_frame.tail(n)</code> will show you the bottom n rows of the data frame.


In [29]:
# show the first 5 rows using dataframe.head() method
print("The first 5 rows of the dataframe") 
df.head(5)

The first 5 rows of the dataframe


0    1            2    3    4     5            6    7      8     9   ...  \
0   3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...   
1   3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...   
2   1    ?  alfa-romero  gas  std   two    hatchback  rwd  front  94.5  ...   
3   2  164         audi  gas  std  four        sedan  fwd  front  99.8  ...   
4   2  164         audi  gas  std  four        sedan  4wd  front  99.4  ...   

    16    17    18    19    20   21    22  23  24     25  
0  130  mpfi  3.47  2.68   9.0  111  5000  21  27  13495  
1  130  mpfi  3.47  2.68   9.0  111  5000  21  27  16500  
2  152  mpfi  2.68  3.47   9.0  154  5000  19  26  16500  
3  109  mpfi  3.19  3.40  10.0  102  5500  24  30  13950  
4  136  mpfi  3.19  3.40   8.0  115  5500  18  22  17450  

[5 rows x 26 columns]

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #1: </h1>
<b>Check the bottom 10 rows of data frame "df".</b>
</div>


In [30]:
print("The last rows of the dataframe\n")
df.tail(10)


The last rows of the dataframe



0    1      2       3      4     5      6    7      8      9   ...   16  \
195  -1   74  volvo     gas    std  four  wagon  rwd  front  104.3  ...  141   
196  -2  103  volvo     gas    std  four  sedan  rwd  front  104.3  ...  141   
197  -1   74  volvo     gas    std  four  wagon  rwd  front  104.3  ...  141   
198  -2  103  volvo     gas  turbo  four  sedan  rwd  front  104.3  ...  130   
199  -1   74  volvo     gas  turbo  four  wagon  rwd  front  104.3  ...  130   
200  -1   95  volvo     gas    std  four  sedan  rwd  front  109.1  ...  141   
201  -1   95  volvo     gas  turbo  four  sedan  rwd  front  109.1  ...  141   
202  -1   95  volvo     gas    std  four  sedan  rwd  front  109.1  ...  173   
203  -1   95  volvo  diesel  turbo  four  sedan  rwd  front  109.1  ...  145   
204  -1   95  volvo     gas  turbo  four  sedan  rwd  front  109.1  ...  141   

       17    18    19    20   21    22  23  24     25  
195  mpfi  3.78  3.15   9.5  114  5400  23  28  13415  
196  mpfi  3.78  3.15   9.5  114  5400  24  28  15985  
197  mpfi  3.78  3.15   9.5  114  5400  24  28  16515  
198  mpfi  3.62  3.15   7.5  162  5100  17  22  18420  
199  mpfi  3.62  3.15   7.5  162  5100  17  22  18950  
200  mpfi  3.78  3.15   9.5  114  5400  23  28  16845  
201  mpfi  3.78  3.15   8.7  160  5300  19  25  19045  
202  mpfi  3.58  2.87   8.8  134  5500  18  23  21485  
203   idi  3.01  3.40  23.0  106  4800  26  27  22470  
204  mpfi  3.78  3.15   9.5  114  5400  19  25  22625  

[10 rows x 26 columns]

<details><summary>Click here for the solution</summary>

```python
print("The last 10 rows of the dataframe\n")
df.tail(10)
```


<h3>Add Headers</h3>
<p>
Take a look at the data set. Pandas automatically set the header with an integer starting from 0.
</p>
<p>
To better describe the data, you can introduce a header. This information is available at:  <a href="https://archive.ics.uci.edu/ml/datasets/Automobile" target="_blank">https://archive.ics.uci.edu/ml/datasets/Automobile</a>.
</p>
<p>
Thus, you have to add headers manually.
</p>
<p>
First, create a list "headers" that include all column names in order.
Then, use <code>dataframe.columns = headers</code> to replace the headers with the list you created.
</p>


In [31]:
# create headers list
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]
print("headers\n", headers)

headers
 ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']


 Replace headers and recheck our data frame:


In [32]:
df.columns = headers
df.columns

Index(['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration',
       'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
       'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
       'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
       'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg',
       'highway-mpg', 'price'],
      dtype='object')

You can also see the first 10 entries of the updated data frame and note that the headers are updated.


In [33]:
df.head(10)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   
5          2                 ?         audi       gas        std          two   
6          1               158         audi       gas        std         four   
7          1                 ?         audi       gas        std         four   
8          1               158         audi       gas      turbo         four   
9          0                 ?         audi       gas      turbo          two   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   
5        sedan          fwd           front        99.8  ...          136   
6        sedan          fwd           front       105.8  ...          136   
7        wagon          fwd           front       105.8  ...          136   
8        sedan          fwd           front       105.8  ...          131   
9    hatchback          4wd           front        99.5  ...          131   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   
5         mpfi  3.19    3.40               8.5        110      5500       19   
6         mpfi  3.19    3.40               8.5        110      5500       19   
7         mpfi  3.19    3.40               8.5        110      5500       19   
8         mpfi  3.13    3.40               8.3        140      5500       17   
9         mpfi  3.13    3.40               7.0        160      5500       16   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  
5          25  15250  
6          25  17710  
7          25  18920  
8          20  23875  
9          22      ?  

[10 rows x 26 columns]

Now, we need to replace the "?" symbol with NaN so the dropna() can remove the missing values:


In [40]:
df1=df.replace('?',np,None)


TypeError: NDFrame.replace() takes from 1 to 3 positional arguments but 4 were given

You can drop missing values along the column "price" as follows:


In [41]:
df=df1.dropna(subset=["price"], axis=0)
df.head(20)

symboling                                  normalized-losses         make  \
0           3  <module 'numpy' from 'C:\\Users\\Junaid Ahamed...  alfa-romero   
1           3  <module 'numpy' from 'C:\\Users\\Junaid Ahamed...  alfa-romero   
2           1  <module 'numpy' from 'C:\\Users\\Junaid Ahamed...  alfa-romero   
3           2                                                164         audi   
4           2                                                164         audi   
5           2  <module 'numpy' from 'C:\\Users\\Junaid Ahamed...         audi   
6           1                                                158         audi   
7           1  <module 'numpy' from 'C:\\Users\\Junaid Ahamed...         audi   
8           1                                                158         audi   
9           0  <module 'numpy' from 'C:\\Users\\Junaid Ahamed...         audi   
10          2                                                192          bmw   
11          0                                                192          bmw   
12          0                                                188          bmw   
13          0                                                188          bmw   
14          1  <module 'numpy' from 'C:\\Users\\Junaid Ahamed...          bmw   
15          0  <module 'numpy' from 'C:\\Users\\Junaid Ahamed...          bmw   
16          0  <module 'numpy' from 'C:\\Users\\Junaid Ahamed...          bmw   
17          0  <module 'numpy' from 'C:\\Users\\Junaid Ahamed...          bmw   
18          2                                                121    chevrolet   
19          1                                                 98    chevrolet   

   fuel-type aspiration num-of-doors   body-style drive-wheels  \
0        gas        std          two  convertible          rwd   
1        gas        std          two  convertible          rwd   
2        gas        std          two    hatchback          rwd   
3        gas        std         four        sedan          fwd   
4        gas        std         four        sedan          4wd   
5        gas        std          two        sedan          fwd   
6        gas        std         four        sedan          fwd   
7        gas        std         four        wagon          fwd   
8        gas      turbo         four        sedan          fwd   
9        gas      turbo          two    hatchback          4wd   
10       gas        std          two        sedan          rwd   
11       gas        std         four        sedan          rwd   
12       gas        std          two        sedan          rwd   
13       gas        std         four        sedan          rwd   
14       gas        std         four        sedan          rwd   
15       gas        std         four        sedan          rwd   
16       gas        std          two        sedan          rwd   
17       gas        std         four        sedan          rwd   
18       gas        std          two    hatchback          fwd   
19       gas        std          two    hatchback          fwd   

   engine-location  wheel-base  ...  engine-size  fuel-system  bore  stroke  \
0            front        88.6  ...          130         mpfi  3.47    2.68   
1            front        88.6  ...          130         mpfi  3.47    2.68   
2            front        94.5  ...          152         mpfi  2.68    3.47   
3            front        99.8  ...          109         mpfi  3.19    3.40   
4            front        99.4  ...          136         mpfi  3.19    3.40   
5            front        99.8  ...          136         mpfi  3.19    3.40   
6            front       105.8  ...          136         mpfi  3.19    3.40   
7            front       105.8  ...          136         mpfi  3.19    3.40   
8            front       105.8  ...          131         mpfi  3.13    3.40   
9            front        99.5  ...          131         mpfi  3.13    3.40   
10           front       101.2  ...          1

Here, `axis=0` means that the contents along the entire row will be dropped wherever the entity 'price' is found to be NaN

Now, you have successfully read the raw data set and added the correct headers into the data frame.


 <div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #2: </h1>
<b>Find the name of the columns of the dataframe.</b>
</div>


In [43]:
# Write your code below and press Shift+Enter to execute 
print(df.columns)

Index(['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration',
       'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
       'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
       'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
       'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg',
       'highway-mpg', 'price'],
      dtype='object')


<details><summary>Click here for the solution</summary>

```python
print(df.columns)
```

</details>


<h2>Save Dataset</h2>
<p>
Correspondingly, Pandas enables you to save the data set to CSV. By using the <code>dataframe.to_csv()</code> method, you can add the file path and name along with quotation marks in the brackets.
</p>
<p>
For example, if you save the data frame <b>df</b> as <b>automobile.csv</b> to your local machine, you may use the syntax below, where <code>index = False</code> means the row names will not be written.
</p>


You can also read and save other file formats. You can use similar functions like **`pd.read_csv()`** and **`df.to_csv()`** for other data formats. The functions are listed in the following table:


<h2>Read/Save Other Data Formats</h2>

| Data Formate |        Read       |            Save |
| ------------ | :---------------: | --------------: |
| csv          |  `pd.read_csv()`  |   `df.to_csv()` |
| json         |  `pd.read_json()` |  `df.to_json()` |
| excel        | `pd.read_excel()` | `df.to_excel()` |
| hdf          |  `pd.read_hdf()`  |   `df.to_hdf()` |
| sql          |  `pd.read_sql()`  |   `df.to_sql()` |
| ...          |        ...        |             ... |


# Basic Insights from the Data set
<p>
After reading data into Pandas dataframe, it is time for you to explore the data set.<br>

There are several ways to obtain essential insights of the data to help you better understand it.
</p>


<h2>Data Types</h2>
<p>
Data has a variety of types.<br>

The main types stored in Pandas data frames are <b>object</b>, <b>float</b>, <b>int</b>, <b>bool</b> and <b>datetime64</b>. In order to better learn about each attribute, you should always know the data type of each column. In Pandas:
</p>


In [44]:
df.dtypes


symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

Returns a series with the data type of each column.


In [45]:
# check the data type of data frame "df" by .dtypes
print(df.dtypes)

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object


<p>
As shown above, you can clearly to see that the data type of "symboling" and "curb-weight" are <code>int64</code>, "normalized-losses" is <code>object</code>, and "wheel-base" is <code>float64</code>, etc.
</p>
<p>
These data types can be changed; you will learn how to accomplish this in a later module.
</p>


<h2>Describe</h2>
If we would like to get a statistical summary of each column such as count, column mean value, column standard deviation, etc., use the describe method:


This method will provide various summary statistics, excluding <code>NaN</code> (Not a Number) values.


In [46]:
df.describe()

symboling  wheel-base      length       width      height  \
count  205.000000  205.000000  205.000000  205.000000  205.000000   
mean     0.834146   98.756585  174.049268   65.907805   53.724878   
std      1.245307    6.021776   12.337289    2.145204    2.443522   
min     -2.000000   86.600000  141.100000   60.300000   47.800000   
25%      0.000000   94.500000  166.300000   64.100000   52.000000   
50%      1.000000   97.000000  173.200000   65.500000   54.100000   
75%      2.000000  102.400000  183.100000   66.900000   55.500000   
max      3.000000  120.900000  208.100000   72.300000   59.800000   

       curb-weight  engine-size  compression-ratio    city-mpg  highway-mpg  
count   205.000000   205.000000         205.000000  205.000000   205.000000  
mean   2555.565854   126.907317          10.142537   25.219512    30.751220  
std     520.680204    41.642693           3.972040    6.542142     6.886443  
min    1488.000000    61.000000           7.000000   13.000000    16.000000  
25%    2145.000000    97.000000           8.600000   19.000000    25.000000  
50%    2414.000000   120.000000           9.000000   24.000000    30.000000  
75%    2935.000000   141.000000           9.400000   30.000000    34.000000  
max    4066.000000   326.000000          23.000000   49.000000    54.000000

<p>
This shows the statistical summary of all numeric-typed (int, float) columns.<br>

For example, the attribute "symboling" has 205 counts, the mean value of this column is 0.83, the standard deviation is 1.25, the minimum value is -2, 25th percentile is 0, 50th percentile is 1, 75th percentile is 2, and the maximum value is 3.
<br>

However, what if you would also like to check all the columns including those that are of type object?
<br><br>

You can add an argument <code>include = "all"</code> inside the bracket. Try it again.

</p>


In [47]:
# describe all the columns in "df" 
df.describe(include = "all")

symboling                                  normalized-losses    make  \
count   205.000000                                                205     205   
unique         NaN                                                 52      22   
top            NaN  <module 'numpy' from 'C:\\Users\\Junaid Ahamed...  toyota   
freq           NaN                                                 41      32   
mean      0.834146                                                NaN     NaN   
std       1.245307                                                NaN     NaN   
min      -2.000000                                                NaN     NaN   
25%       0.000000                                                NaN     NaN   
50%       1.000000                                                NaN     NaN   
75%       2.000000                                                NaN     NaN   
max       3.000000                                                NaN     NaN   

       fuel-type aspiration num-of-doors body-style drive-wheels  \
count        205        205          205        205          205   
unique         2          2            3          5            3   
top          gas        std         four      sedan          fwd   
freq         185        168          114         96          120   
mean         NaN        NaN          NaN        NaN          NaN   
std          NaN        NaN          NaN        NaN          NaN   
min          NaN        NaN          NaN        NaN          NaN   
25%          NaN        NaN          NaN        NaN          NaN   
50%          NaN        NaN          NaN        NaN          NaN   
75%          NaN        NaN          NaN        NaN          NaN   
max          NaN        NaN          NaN        NaN          NaN   

       engine-location  wheel-base  ...  engine-size  fuel-system  bore  \
count              205  205.000000  ...   205.000000          205   205   
unique               2         NaN  ...          NaN            8    39   
top              front         NaN  ...          NaN         mpfi  3.62   
freq               202         NaN  ...          NaN           94    23   
mean               NaN   98.756585  ...   126.907317          NaN   NaN   
std                NaN    6.021776  ...    41.642693          NaN   NaN   
min                NaN   86.600000  ...    61.000000          NaN   NaN   
25%                NaN   94.500000  ...    97.000000          NaN   NaN   
50%                NaN   97.000000  ...   120.000000          NaN   NaN   
75%                NaN  102.400000  ...   141.000000          NaN   NaN   
max                NaN  120.900000  ...   326.000000          NaN   NaN   

        stroke compression-ratio horsepower  peak-rpm    city-mpg highway-mpg  \
count      205        205.000000        205       205  205.000000  205.000000   
unique      37               NaN         60        24         NaN         NaN   
top       3.40               NaN         68      5500         NaN         NaN   
freq        20               NaN         19        37         NaN         NaN   
mean       NaN         10.142537        NaN       NaN   25.219512   30.751220   
std        NaN          3.972040        NaN       NaN    6.542142    6.886443   
min        NaN          7.000000        NaN       NaN   13.000000   16.000000   
25%        NaN          8.600000        NaN       NaN   19.000000   25.000000   
50%        NaN          9.000000        NaN       NaN   24.000000   30.000000   
75%        NaN          9.400000        NaN       NaN   30.000000   34.000000   
max        NaN         23.000000        NaN       NaN   49.000000   54.000000   

                                                    price  
count                                                 205  
unique                                                187  
top     <module 'numpy' from 'C:\\Users\\Junaid Ahamed...  
freq                                                    4  
mean                                         

<p>
Now it provides the statistical summary of all the columns, including object-typed attributes.<br>

YOu can now see how many unique values there, which one is the top value, and the frequency of the top value in the object-typed columns.<br>

Some values in the table above show "NaN". Those numbers are not available regarding a particular column type.<br>
</p>


<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #3: </h1>

<p>
You can select the columns of a dataframe by indicating the name of each column. For example, you can select the three columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3']]</code>
</p>
<p>
Where "column" is the name of the column, you can apply the method  ".describe()" to get the statistics of those columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3'] ].describe()</code>
</p>

Apply the  method to ".describe()" to the columns 'length' and 'compression-ratio'.

</div>


In [48]:
# Write your code below and press Shift+Enter to execute 
df[['length','compression-ratio']].describe()

length  compression-ratio
count  205.000000         205.000000
mean   174.049268          10.142537
std     12.337289           3.972040
min    141.100000           7.000000
25%    166.300000           8.600000
50%    173.200000           9.000000
75%    183.100000           9.400000
max    208.100000          23.000000


<details><summary>Click here for the solution</summary>

```python
df[['length', 'compression-ratio']].describe()
```

</details>


<h2>Info</h2>
You can also use another method to check your data set:


It provides a concise summary of your data frame. 

This method prints information about a data frame including the index dtype and columns, non-null values and memory usage.


In [49]:
# look at the info of "df"
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  205 non-null    object 
 2   make               205 non-null    object 
 3   fuel-type          205 non-null    object 
 4   aspiration         205 non-null    object 
 5   num-of-doors       205 non-null    object 
 6   body-style         205 non-null    object 
 7   drive-wheels       205 non-null    object 
 8   engine-location    205 non-null    object 
 9   wheel-base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb-weight        205 non-null    int64  
 14  engine-type        205 non-null    object 
 15  num-of-cylinders   205 non-null    object 
 16  engine-size        205 non

<h1>Excellent! You have just completed the  Introduction Notebook.</h1>


### Thank you for completing this lab.

## Author

<a href="https://www.linkedin.com/in/joseph-s-50398b136/" target="_blank">Joseph Santarcangelo</a>

### Other Contributors

<a href="https://www.linkedin.com/in/mahdi-noorian-58219234/" target="_blank">Mahdi Noorian PhD</a>

Bahare Talayian

Eric Xiao

Steven Dong

Parizad

Hima Vasudevan

<a href="https://www.linkedin.com/in/fiorellawever/" target="_blank">Fiorella Wenver</a>

<a href=" https://www.linkedin.com/in/yi-leng-yao-84451275/ " target="_blank" >Yi Yao</a>.

<a href="https://www.coursera.org/instructor/~129186572/ " target="_blank" >Abhishek Gagneja</a>

# <h3 align="center"> © IBM Corporation 2023. All rights reserved. <h3/>

<!--## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description                       |
| ----------------- | ------- | ---------- | ---------------------------------------- |
| 2023-09-28        | 2.4     | Abhishek Gagneja| Minor instructional update          |
| 2020-10-30        | 2.3     | Lakshmi    | Changed URL of the csv                   |
| 2020-09-22        | 2.2     | Nayef      | Added replace() method to remove '?'     |
| 2020-09-09        | 2.1     | Lakshmi    | Made changes in info method of dataframe |
| 2020-08-27        | 2.0     | Lavanya    | Moved lab to course repo in GitLab       |


--!> 
<hr>
